---
title: "Gene Expression Response to Influenza Vaccination"
author: "Kaitlyn Cross, Eunchong Kang, Yu (Sylvia) Zhang"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
  wrap: 80
---

```{r load library, message = FALSE}
library(ggplot2)
library(dplyr)
library(rmarkdown)
library(knitr)
library(GEOquery)
library(SummarizedExperiment)
library(DESeq2)
library(here)
library(limma)
library(nlme)
library(MASS)
library(multcomp)
library(pheatmap)
library(variancePartition)
library(lme4)
library("r2glmm")
library(mclust)
```

# Introduction

# Data collection

Data for the study on trivalent influenza vaccine are deposited in National Center for Biotechnology Information (NCBI) Gene Expression Omnibus (GEO) with accession GSE48024. It was published in 2013. 

A total of 119 male and 128 female between 19 to 41 years of age were recruited in the study. Whole blood samples were drawn before vaccination (day 0) and at three time points after vaccination (day 1, 3 and 14). All male samples are sequenced using Illumina HumanHT-12 V3.0 expression beadchip and all female samples are sequence using Illumina HumanHT-12 V4.0 expression beadchip. 

### Load data

Two separate datasets are loaded, one for female cohort and another one for male cohort.

```{r}
gset <- getGEO("GSE48024", GSEMatrix =TRUE, AnnotGPL=TRUE)

gset_female <- gset[[1]]
gset_male <- gset[[2]]

# Check gender 
unique(gset_female$`gender:ch1`) 
# Check gender 
unique(gset_male$`gender:ch1`)

# Check sequencing platform 
unique(gset_female$platform_id) #Illumina HumanHT-12 V4.0 expression beadchip
# Check sequencing platform 
unique(gset_male$platform_id) #Illumina HumanHT-12 V3.0 expression beadchip

```

Based on the data loaded directly from GEO, 116 out of 119 males and 110 out of 128 females are available. A total of 417 samples are from female cohort and 431 samples are from male cohort.

```{r}
# Check number of subjects available in data
length(unique(gset_female$`subject:ch1`))
# Check number of subjects available in data
length(unique(gset_male$`subject:ch1`))

# Check number of samples available in data
ncol(gset_female)
# Check number of samples available in data
ncol(gset_male)
```

There are 47276 probes used for female cohort, and 48742 probes used for male cohort.

```{r}
# Check number of probes
nrow(gset_female)
# Check number of probes
nrow(gset_male)
```

Data has been processed and normalized according to description given in GEO. We randomly select 10 samples from each cohort to check the distribution of normalized data.

```{r}
# Data processing + normalization 
unique(gset_male$data_processing)
# Data processing + normalization 
unique(gset_female$data_processing)

boxplot(exprs(gset_male)[,sample(1:431, 10, replace=F)], 
        main = "Ten random samples from male cohort", ylab = "Expression")

boxplot(exprs(gset_female)[,sample(1:417, 10, replace=F)], 
        main = "Ten random samples from female cohort", ylab = "Expression")

```

Last, we check how many samples at each sampling time for each cohort.

```{r}
# For female
table(gset_female$`time:ch1`)

# For male
table(gset_male$`time:ch1`)

```

Because female and male cohorts are sequenced using different platforms, all further data processing and analysis will be done separately. 

# Analysis plan

# Pre-processing

First to setup SummarizedExperiment object.

```{r}
SE_male  <- SummarizedExperiment(assay = list("exprs" = exprs(gset_male)),
                                  colData = as(gset_male@phenoData, "data.frame"),
                                  rowData = as(gset_male@featureData, "data.frame"))

SE_female  <- SummarizedExperiment(assay = list("exprs" = exprs(gset_female)),
                                  colData = as(gset_female@phenoData, "data.frame"),
                                  rowData = as(gset_female@featureData, "data.frame"))

# Create few variables with better names 
SE_male$time <- factor(SE_male$`time:ch1`, levels = c("Day0", "Day1", "Day3", "Day14"))
SE_male$ID <- SE_male$`subject:ch1`
SE_female$time <- factor(SE_female$`time:ch1`, levels = c("Day0", "Day1", "Day3", "Day14"))
SE_female$ID <- SE_female$`subject:ch1`

```

### Filter samples

As we have seen in Data Collection section, not all participants are sampled at all four time points. Thus, we keep those who sampled at all time. 

```{r}
male.all <- as.data.frame(table(SE_male$ID)) %>% filter(Freq == 4) %>% rename( "Var1" = "ID")
female.all <- as.data.frame(table(SE_female$ID)) %>% filter(Freq == 4) %>% rename( "Var1" = "ID")

SE_male <- SE_male[, which(SE_male$ID %in% male.all$ID)]
SE_female <- SE_female[, which(SE_female$ID %in% female.all$ID)]
```

There are 92 males and 87 females sampled at all four time points. These will be used for downstream analysis. 

```{r}
table(SE_male$time)

table(SE_female$time)
```

### Filter probes/genes

#### Step 1: We first subset to probes with corresponding gene symbols.

Probes for male cohort: 48742 -> 29228 

Probes for female cohort: 47276 -> 31243

```{r}

gset_male <- gset_male[which(gset_male@featureData@data[["Gene symbol"]] != ""), ]
nrow(gset_male)

gset_female <- gset_female[which(gset_female@featureData@data[["Gene symbol"]] != ""), ]
nrow(gset_female)



```

#### Step 2: Average expression for probes correponding to the same gene name. 

Using limma() package.

Probes for male cohort: 29228 -> 19590

Probes for female cohort: 31243 -> 20751

````{r}
dupRM_male <-  avereps(assays(SE_male)$exprs, ID=rowData(SE_male)$"Gene symbol")
nrow(dupRM_male)

dupRM_female <-  avereps(assays(SE_female)$exprs, ID=rowData(SE_female)$"Gene symbol")
nrow(dupRM_female)

```

Update SummarizeExperiment object with updated expression matrix.

```{r}
# Make sure samples are ordered the same
all.equal(colnames(dupRM_male) , rownames(SE_male@colData))

SE_male_dupRM <- SummarizedExperiment(assay = list("exprs" = dupRM_male),
                                  colData = SE_male@colData)

all.equal(colnames(dupRM_female) , rownames(SE_female@colData))

SE_female_dupRM <- SummarizedExperiment(assay = list("exprs" = dupRM_female),
                                  colData = SE_female@colData)

```

#### Step 3: Keep top 75% most variance probes/genes for efficient downstream analysis.

Probes for male cohort: 19590 -> 14692

Probes for female cohort: 20751 -> 15563

```{r}
rowData(SE_male_dupRM)$gene_variance <- rowVars(assays(SE_male_dupRM)$exprs)
quantile(rowData(SE_male_dupRM)$gene_variance)

SE_male_dupRM_25up <- SE_male_dupRM[which(rowData(SE_male_dupRM)$gene_variance > 0.0022386139), ]

rowData(SE_female_dupRM)$gene_variance <- rowVars(assays(SE_female_dupRM)$exprs)
quantile(rowData(SE_female_dupRM)$gene_variance)

SE_female_dupRM_25up <- SE_female_dupRM[which(rowData(SE_female_dupRM)$gene_variance > 0.0070110061), ]

```
For downstream analysis, we use 14692 unique genes and 368 samples from the male cohort, and 15563 genes with 348 samples from the female cohort.

```{r}
SE_male_dupRM_25up

SE_female_dupRM_25up

```

### GMM clustering to infer response to vaccination

Data on antibody titer are deposited on dbGaP with restricted access. Thus, we use 11 genes given by authors to differentiate high response and low response after vaccination. These 11 genes are found using male cohort.

Gene PRDX2 are not present in both female and male cohort after data pre-processing.

```{r}
DE11_response <- c("STAT1", "IRF9", "SPI1", "CD74", "HLA-E", "TNFSF13B", "PRDX2", "PRDX3", "E2F2", "PTEN", "ITGB1")

DE11_response[!DE11_response %in% rownames(SE_male_dupRM_25up)]

DE11_response[!DE11_response %in% rownames(SE_female_dupRM_25up)]
```

#### On male cohort

```{r}
male_exprs_day1_0 <- list( day1 = SE_male_dupRM_25up[which(rownames(SE_male_dupRM_25up) %in% DE11_response),which(SE_male_dupRM_25up$time == "Day1")],
                           day0 = SE_male_dupRM_25up[which(rownames(SE_male_dupRM_25up) %in% DE11_response),which(SE_male_dupRM_25up$time == "Day0")])
colnames(male_exprs_day1_0[["day1"]]) <- male_exprs_day1_0[["day1"]]$ID
colnames(male_exprs_day1_0[["day0"]]) <- male_exprs_day1_0[["day0"]]$ID

male_exprs_day1_0[["day1"]] <- as.matrix(assay(male_exprs_day1_0[["day1"]]))
male_exprs_day1_0[["day0"]] <- as.matrix(assay(male_exprs_day1_0[["day0"]]))

dim(male_exprs_day1_0[[1]])

dim(male_exprs_day1_0[[2]])

# Match the order of gene and order of participants 
male_exprs_day1_0[[2]] <- male_exprs_day1_0[[2]][match(rownames(male_exprs_day1_0[[1]]), rownames(male_exprs_day1_0[[2]])), match(colnames(male_exprs_day1_0[[1]]), colnames(male_exprs_day1_0[[2]]))]

# Subtract day 0 from day 1 for each participants
male_diff_day1_0 <- (male_exprs_day1_0[["day1"]] - male_exprs_day1_0[["day0"]])

male_diff_day1_0[,1:5]

# GMM cluster
GMM_male <-  Mclust(t(male_diff_day1_0),  G = 2)
table(GMM_male$classification)
```

According to authors, genes ("STAT1", "IRF9", "SPI1", "CD74", "HLA-E", "TNFSF13B") are up-regulated in high responders, and remaining genes in the list are up-regulated in low responders. 

We assign group 2 as high responders based on boxplots on the difference of expression between day 1 and day 0 by clusters from GMM. 

```{r, fig.width= 8, fig.height=5}

# Check GMM cluster result wrt 10 genes
par(mfrow=c(2,5))
for (i in 1:nrow(male_diff_day1_0)){
  dff <- data.frame(expr = male_diff_day1_0[rownames(male_diff_day1_0)[i],], group=GMM_male$classification)
  boxplot(expr ~ group,dff, main=rownames(male_diff_day1_0)[i], ylab="diff(day1- day0)")
}

```

```{r}
SE_male_dupRM_25up$response <- factor(GMM_male$classification[match(SE_male_dupRM_25up$ID, names(GMM_male$classification))],  levels = c(1,2))
```

#### On male cohort

Please refer to the Rmd file for code. 

```{r, echo = FALSE, message = FALSE}
female_exprs_day1_0 <- list( day1 = SE_female_dupRM_25up[which(rownames(SE_female_dupRM_25up) %in% DE11_response),which(SE_female_dupRM_25up$time == "Day1")],
                           day0 = SE_female_dupRM_25up[which(rownames(SE_female_dupRM_25up) %in% DE11_response),which(SE_female_dupRM_25up$time == "Day0")])
colnames(female_exprs_day1_0[["day1"]]) <- female_exprs_day1_0[["day1"]]$ID
colnames(female_exprs_day1_0[["day0"]]) <- female_exprs_day1_0[["day0"]]$ID

female_exprs_day1_0[["day1"]] <- as.matrix(assay(female_exprs_day1_0[["day1"]]))
female_exprs_day1_0[["day0"]] <- as.matrix(assay(female_exprs_day1_0[["day0"]]))

#dim(female_exprs_day1_0[[1]])

#dim(female_exprs_day1_0[[2]])

# Match the order of gene and order of participants 
female_exprs_day1_0[[2]] <- female_exprs_day1_0[[2]][match(rownames(female_exprs_day1_0[[1]]), rownames(female_exprs_day1_0[[2]])), match(colnames(female_exprs_day1_0[[1]]), colnames(female_exprs_day1_0[[2]]))]

# Subtract day 0 from day 1 for each participants
female_diff_day1_0 <- (female_exprs_day1_0[["day1"]] - female_exprs_day1_0[["day0"]])
```

GMM result.

```{r, echo = FALSE}
# GMM cluster
GMM_female <-  Mclust(t(female_diff_day1_0),  G = 2)
table(GMM_female$classification)
```

The 10 genes used for clustering are found using male cohort, and analysis using GMM shows inconsistencies when applied on female cohort. Overall, we assign group 1 as high responders and group 2 as low responders for female cohort. 

```{r, fig.width= 8, fig.height=5, echo = FALSE}

# Check GMM cluster result wrt 10 genes
par(mfrow=c(2,5))
for (i in 1:nrow(female_diff_day1_0)){
  dff <- data.frame(expr = female_diff_day1_0[rownames(female_diff_day1_0)[i],], group=GMM_female$classification)
  boxplot(expr ~ group,dff, main=rownames(female_diff_day1_0)[i], ylab="diff(day1- day0)")
}

```

```{r, echo = FALSE}
SE_female_dupRM_25up$response <- factor(GMM_female$classification[match(SE_female_dupRM_25up$ID, names(GMM_female$classification))],  levels = c(2,1))
```




# Results

We first define functions for analysis.

```{r, eval = FALSE}
mixed_test <- function(gene,gene_name, covariates, fml, var_rowname, j ){

  df_test <- cbind(gene , covariates)
  colnames(df_test)[1] <- "gene"
  mix_model <- aov(fml, data=df_test)
  df_ret <- data.frame(unlist(summary(mix_model))[paste0("Error: Within.Pr(>F)",j)])
  rownames(df_ret) <- var_rowname
  colnames(df_ret) <- gene_name
  return(tibble(df_ret))
}

mixed_test2 <- function(gene,gene_name, covariates, fml, var_rowname){
  
   #gene = as.data.frame(SE_male_dupRM_25up@assays@data@listData[["exprs"]][which(rownames(SE_male_dupRM_25up) ==  genes[1]),])
  #gene_name = genes[1]
  #covariates = cov
  #fml = as.formula(gene ~ time*response + Error(ID))
  #var_rowname = c("response_time")
  
  df_test <- cbind(gene , covariates)
  colnames(df_test)[1] <- "gene"
  mix_model <- aov(fml, data=df_test)
  df_ret <- data.frame(unlist(summary(mix_model))["Error: Within.Pr(>F)2"])
  rownames(df_ret) <- var_rowname
  colnames(df_ret) <- gene_name
  return(tibble(df_ret))
}

```

## Male cohort

### Differential expression analysis wrt time

We first fit mixed effect model, then extract p-value of testing the effect on time variable which is treated as categorical variable.

```{r, eval = FALSE}

cov_male <- data.frame(ID = SE_male_dupRM_25up$ID,
                  time = SE_male_dupRM_25up$time)

genes_male <- rownames(assays(SE_female_dupRM_25up)$exprs)
mix_test_time_male <- apply(assays(SE_female_dupRM_25up)$exprs, 1,
                            function(gene_i) mixed_test( gene = gene_i,
                                                         gene_name = genes_male[i],
                                                         covariates = cov_male,
                                                         fml = as.formula(gene ~ time + Error(ID)),
                                                         var_rowname = c("time")) )
  

```

#### GO analysis

# Conclusion and discussion



