---
title: "Gene Expression Response to Influenza Vaccination"
author: "Kaitlyn Cross, Eunchong Kang, Yu (Sylvia) Zhang"
date: "`r Sys.Date()`"
output:   
  html_document:
        toc: true
        toc_float: true
editor_options: 
  markdown: 
    wrap: 72
  wrap: 80
---

```{r load library, message = FALSE}
library(ggplot2)
library(dplyr)
library(rmarkdown)
library(knitr)
library(GEOquery)
library(SummarizedExperiment)
library(DESeq2)
library(here)
library(limma)
library(nlme)
library(MASS)
library(multcomp)
library(pheatmap)
library(variancePartition)
library(lme4)
library("r2glmm")
library(mclust)
library(gprofiler2)
library(enrichplot)
```

# Introduction

The development of effective vaccines to prevent the spread of infectious disease is one of the greatest public 
health achievements of the 20th century. While effective vaccines are available for most widespread infectious
diseases, the exact efficacy of the vaccines in terms of individual protection against acquiring disease are
variable based on the disease itself and the vaccine platform. A strong area of interest in vaccine development
is identifying technologies and strategies for increasing the strength and duration of immune response to vaccination. Influenza vaccines in particular are known to have limited efficacy and duration, and developing a
long-lasting universal influenza vaccine is a top target of the NIH's National Institute for Allergy and Infectious
Diseases (NIAID).

One way to measure a vaccine's impact in an individual is to measure specific immune products, called correlates of protection (CoP), that are known indicators of a vaccine's efficacy. In influenza vaccines, common CoPs are 
Hemagglutination Inhibition (HAI) and Neutralizing (Neut) antibodies. These antibodies are the end products of the
complex immune response inititated at the vaccination event. One goal among vaccine experts currently is to better
understand the underlying differences between individuals who have a robust immune response to a particular vaccine (high antibody titers, indicating low likelihood of infection) versus those who have little to no response to the same stimulus. One way to examine for the underlying differences in biological processes is to take a global look
at gene expression in response to vaccination, and to identify genes or pathways that are differentially expressed between high and low responders. This is the approach that Bucasas et. al. (2011) took. In our project, we will replicate
the analysis approach using data accessed from the National Center for Biotechnology Information (NCBI) Gene Expression Omnibus (GEO).

# Data collection

Data for the study on trivalent influenza vaccine is available in GEO with accession GSE48024. The manuscript describing results found by the study group in the male cohort was published in 2011. Data was published to the NCBI GEO in 2013. 

A total of 119 males and 128 females between 19 to 41 years of age were recruited in the study and vaccinated with trivalent influenza vaccine. Whole blood samples were drawn before vaccination (day 0) and at three time points after vaccination (day 1, 3 and 14). All male samples are sequenced using Illumina HumanHT-12 V3.0 expression beadchip and all female samples are sequence using Illumina HumanHT-12 V4.0 expression beadchip. 

### Load data

Two separate datasets are loaded, one for female cohort and another one for male cohort.

```{r}
gset <- getGEO("GSE48024", GSEMatrix =TRUE, AnnotGPL=TRUE)

gset_female <- gset[[1]]
gset_male <- gset[[2]]

# Check gender 
unique(gset_female$`gender:ch1`) 
# Check gender 
unique(gset_male$`gender:ch1`)

# Check sequencing platform 
unique(gset_female$platform_id) #Illumina HumanHT-12 V4.0 expression beadchip
# Check sequencing platform 
unique(gset_male$platform_id) #Illumina HumanHT-12 V3.0 expression beadchip

```

Based on the data loaded directly from GEO, 116 out of 119 males and 110 out of 128 females are available. A total of 417 samples are from female cohort and 431 samples are from male cohort.

```{r}
# Check number of subjects available in data
length(unique(gset_female$`subject:ch1`))
# Check number of subjects available in data
length(unique(gset_male$`subject:ch1`))

# Check number of samples available in data
ncol(gset_female)
# Check number of samples available in data
ncol(gset_male)
```

There are 47276 probes used for female cohort, and 48742 probes used for male cohort.

```{r}
# Check number of probes
nrow(gset_female)
# Check number of probes
nrow(gset_male)
```

Data has been processed and normalized according to description given in GEO. We randomly select 10 samples from each cohort to check the distribution of normalized data.

```{r}
# Data processing + normalization 
unique(gset_male$data_processing)
# Data processing + normalization 
unique(gset_female$data_processing)

boxplot(exprs(gset_male)[,sample(1:431, 10, replace=F)], 
        main = "Ten random samples from male cohort", ylab = "Expression")

boxplot(exprs(gset_female)[,sample(1:417, 10, replace=F)], 
        main = "Ten random samples from female cohort", ylab = "Expression")

```

Last, we check how many samples at each sampling time for each cohort.

```{r}
# For female
table(gset_female$`time:ch1`)

# For male
table(gset_male$`time:ch1`)

```

Because female and male cohorts are sequenced using different platforms, all further data processing and analysis will be done separately. 

# Analysis plan

## Statistical Model

We use repeated measures ANOVA (or Mixed effect ANOVA) to implement statistical analysis. It is suitable for a well-balanced data with correlation in observations. The advantage of the model is that it does not need an assumption of a linearity or any curve so that we do not need any consideration on the shape of trajectories between time points. Also, it is impossible to check linearity assumption for all over 14,000 genes. The disadvantage of this model is that we can't say about the trend. For example, the model and its result cannot provide any inference on how gene expression level changes between day 3 and day 14.

## DE genes across Time

To examine gene expression across time before and after vaccination, mixed effect models are fitted on each gene with participants as random effect and time as fix effect. Further, time will be analyzed as a categorical variable. We will select participants who are sampled at all time points to have a balanced data, and use compound symmetry covariance structure for mixed effect model. Specifically, the mixed effect model is

$$ expr_{ij} = \beta_0 + \beta_1I(time_{ij} = 1) + \beta_2I(time_{ij} = 3) + \beta_3I(time_{ij} = 14) + b_i + \varepsilon_{ij},$$
where $b_i \sim N(0, \sigma^2_b)$ and $\varepsilon_{ij} \sim N(0, \sigma_e^2 I_4)$ and $b_i \perp \!\!\! \perp \varepsilon_{ij}$.

To determine differentially expressed (DE) genes across time, we use Wald test on
$$
H_0: \beta_1 = \beta_2 = \beta_3 = 0, \hspace{1cm} vs.  \hspace{1cm} H_\alpha: \beta_1 \neq 0 \text{ or } \beta_2 \neq 0 \text{ or } \beta_3 \neq 0.
$$
Essentially, this test detects genes showing different expressions in at least one time point among day 0, 1, 3 and 14. P-values from Wald test are adjusted using Benjamini-Hotchberg (BH) method, and false discovery rate is controlled at 1%. Then Gene Ontology (GO) enrichment analysis will be conducted on the DE genes. 

## DE genes with respect to Response across Time

Next, we are interested in genes differentially expressed with respect to response to vaccination across time. Information on antibody titer is not publicly available, hence response to vaccination is inferred for each paticipant using differentially expressed genes with respect to response found and reported by authors. Gaussian mixture model (GMM) will be used to cluster participants into high responder group and low responder group. Difference between day 1 and day 0 of gene expression is used for GMM. Group assignment is based on visualization of the difference in gene expression. 

After response to vaccination is obtained, mixed effect model is fitted on each gene as shown below,
$$
\begin{align}
    expr_{ij} =\, & \beta_0 + \beta_1I(time_{ij}=1) + \beta_2I(time_{ij}=3) + \beta_3I(time_{ij}=14) + \beta_4I(response_{i} = high)\\
    &+\beta_5I(time_{ij}=1)I(response_{i} = high) + \beta_6I(time_{ij}=3)I(response_{i} = high) \\
    &+ \beta_7I(time_{ij}=14)I(response_{i} = high) + b_i + e_{ij}
\end{align}
$$
where $i$ indicates subject and $j$ does time, $b_i \sim N(0, \sigma^2_b)$, $e_i \sim MVN(0, \sigma^2_eI_4)$ and $b_i \bot e_{ij}$.

The response variable inferred from GMM and its interaction with time variable are added to the model as a categorical variable. The test is about whether the interaction term is important in the model and detects genes showing the unequal difference between two response groups at-least one time point among day 0, 1, 3, and 14. Then, we extract p-value from the test. And, the hypothesis is $H_0: \beta_5 = \beta_6 = \beta_7 = 0$ and $H_a: \beta_5 \neq  0 \,or\, \beta_6 \neq 0 \,or\, \beta_7 \neq 0$.

P-values from Wald test are adjusted using Benjamini-Hotchberg (BH) method, and false discovery rate is controlled at 5%. Then Gene Ontology (GO) enrichment analysis will be conducted on the DE genes. 

However, this analysis has a problem, which is it may detect the genes used clustering and its related genes with high significance because grouping is created by the difference of day 1 and day 0 in the genes listed in the pre-process. Therefore, the result should be carefully interpreted for this analysis.


# Pre-processing

First to setup SummarizedExperiment object:

```{r}
SE_male  <- SummarizedExperiment(assay = list("exprs" = exprs(gset_male)),
                                  colData = as(gset_male@phenoData, "data.frame"),
                                  rowData = as(gset_male@featureData, "data.frame"))

SE_female  <- SummarizedExperiment(assay = list("exprs" = exprs(gset_female)),
                                  colData = as(gset_female@phenoData, "data.frame"),
                                  rowData = as(gset_female@featureData, "data.frame"))

# Create few variables with better names 
SE_male$time <- factor(SE_male$`time:ch1`, levels = c("Day0", "Day1", "Day3", "Day14"))
SE_male$ID <- SE_male$`subject:ch1`
SE_female$time <- factor(SE_female$`time:ch1`, levels = c("Day0", "Day1", "Day3", "Day14"))
SE_female$ID <- SE_female$`subject:ch1`

```

### Filter samples

As shown in Data Collection section, not all participants are sampled at all four time points. Thus, we keep those who sampled at all time. 

```{r}
male.all <- as.data.frame(table(SE_male$ID)) %>% filter(Freq == 4) %>% rename( "Var1" = "ID")
female.all <- as.data.frame(table(SE_female$ID)) %>% filter(Freq == 4) %>% rename( "Var1" = "ID")

SE_male <- SE_male[, which(SE_male$ID %in% male.all$ID)]
SE_female <- SE_female[, which(SE_female$ID %in% female.all$ID)]
```

There are 92 males and 87 females sampled at all four time points. These will be used for downstream analysis. 

```{r}
table(SE_male$time)

table(SE_female$time)
```

### Filter probes/genes

#### Step 1: Subset to probes with corresponding gene symbols.

Probes for male cohort reduces from 48742 to 29228. 

Probes for female cohort reduces from 47276 to 31243.

```{r}

gset_male <- gset_male[which(gset_male@featureData@data[["Gene symbol"]] != ""), ]
nrow(gset_male)

gset_female <- gset_female[which(gset_female@featureData@data[["Gene symbol"]] != ""), ]
nrow(gset_female)



```

#### Step 2: Average expression for probes correponding to the same gene name. 

Using limma() package.

Probes for male cohort reduces from 29228 to 19590.

Probes for female cohort reduces from 31243 to 20751.

```{r}
dupRM_male <-  avereps(assays(SE_male)$exprs, ID=rowData(SE_male)$"Gene symbol")
nrow(dupRM_male)

dupRM_female <-  avereps(assays(SE_female)$exprs, ID=rowData(SE_female)$"Gene symbol")
nrow(dupRM_female)

```

Update SummarizeExperiment object with expression matrix where duplicated genes' expression are averaged.

```{r}
# Make sure samples are ordered the same
all.equal(colnames(dupRM_male) , rownames(SE_male@colData))

SE_male_dupRM <- SummarizedExperiment(assay = list("exprs" = dupRM_male),
                                  colData = SE_male@colData)

all.equal(colnames(dupRM_female) , rownames(SE_female@colData))

SE_female_dupRM <- SummarizedExperiment(assay = list("exprs" = dupRM_female),
                                  colData = SE_female@colData)

```

#### Step 3: Keep top 75% most variant probes/genes for efficient downstream analysis.

Probes for male cohort reduces from 19590 to 14692.

Probes for female cohort from 20751 to 15563.

```{r}
rowData(SE_male_dupRM)$gene_variance <- rowVars(assays(SE_male_dupRM)$exprs)
quantile(rowData(SE_male_dupRM)$gene_variance)

SE_male_dupRM_25up <- SE_male_dupRM[which(rowData(SE_male_dupRM)$gene_variance > 0.0022386139), ]

rowData(SE_female_dupRM)$gene_variance <- rowVars(assays(SE_female_dupRM)$exprs)
quantile(rowData(SE_female_dupRM)$gene_variance)

SE_female_dupRM_25up <- SE_female_dupRM[which(rowData(SE_female_dupRM)$gene_variance > 0.0070110061), ]

```
For downstream analysis, we use 14692 unique genes and 368 samples from the male cohort, and 15563 genes with 348 samples from the female cohort.

```{r}
SE_male_dupRM_25up

SE_female_dupRM_25up

```

### GMM clustering to infer response to vaccination

Data regarding informaiton on antibody titer are deposited in dbGaP with restricted access. Thus, we use 11 differentially expressed genes with respect to response given by authors to differentiate high response and low response after vaccination among participants. Please not that these 11 genes are found using male cohort.

Gene PRDX2 are not present in both female and male cohort after data pre-processing.

```{r}
DE11_response <- c("STAT1", "IRF9", "SPI1", "CD74", "HLA-E", "TNFSF13B", "PRDX2", "PRDX3", "E2F2", "PTEN", "ITGB1")

DE11_response[!DE11_response %in% rownames(SE_male_dupRM_25up)]

DE11_response[!DE11_response %in% rownames(SE_female_dupRM_25up)]
```

We will cluster participants based on the difference between day 1 and day 0 expression of the available 10 genes.

#### On male cohort

```{r}
male_exprs_day1_0 <- list( day1 = SE_male_dupRM_25up[which(rownames(SE_male_dupRM_25up) %in% DE11_response),which(SE_male_dupRM_25up$time == "Day1")],
                           day0 = SE_male_dupRM_25up[which(rownames(SE_male_dupRM_25up) %in% DE11_response),which(SE_male_dupRM_25up$time == "Day0")])
colnames(male_exprs_day1_0[["day1"]]) <- male_exprs_day1_0[["day1"]]$ID
colnames(male_exprs_day1_0[["day0"]]) <- male_exprs_day1_0[["day0"]]$ID

male_exprs_day1_0[["day1"]] <- as.matrix(assay(male_exprs_day1_0[["day1"]]))
male_exprs_day1_0[["day0"]] <- as.matrix(assay(male_exprs_day1_0[["day0"]]))

dim(male_exprs_day1_0[[1]])

dim(male_exprs_day1_0[[2]])

# Match the order of gene and order of participants 
male_exprs_day1_0[[2]] <- male_exprs_day1_0[[2]][match(rownames(male_exprs_day1_0[[1]]), rownames(male_exprs_day1_0[[2]])), match(colnames(male_exprs_day1_0[[1]]), colnames(male_exprs_day1_0[[2]]))]

# Subtract day 0 from day 1 for each participants
male_diff_day1_0 <- (male_exprs_day1_0[["day1"]] - male_exprs_day1_0[["day0"]])

male_diff_day1_0[,1:5]

# GMM cluster
GMM_male <-  Mclust(t(male_diff_day1_0),  G = 2)
table(GMM_male$classification)
```

According to authors, genes ("STAT1", "IRF9", "SPI1", "CD74", "HLA-E", "TNFSF13B") are up-regulated in high responders, and remaining genes in the list are up-regulated in low responders. 

We assign group 2 as high responders based on boxplots on the difference of expression between day 1 and day 0 by clusters from GMM. 

```{r, fig.width= 8, fig.height=5}

# Gene names in Table 1
names_table1 <- c("STAT1", "IRF9", "SPI1", "CD74", "HLA-E", "TNFSF13B", "PRDX3", "E2F2", "PTEN", "ITGB1")

# Create Table1
table1 <- data.frame(Gene=names_table1, "Upregulated in"=c("High responders", "High responders","High responders","High responders","High responders","High responders","Low responders","Low responders","Low responders","Low responders"))

## Subset of Transcripts Used for prediction of the TRI
table1

# Boxplot of diff(day1-day0) by clustered group across the genes in Table 1
par(mfrow=c(2,5))
for (i in 1:length(names_table1)){
  dff <- data.frame(expr = male_diff_day1_0[names_table1[i],], group=GMM_male$classification)
  boxplot(expr ~group,dff, main=names_table1[i], ylab="diff(day1- day0)")
}
par(mfrow=c(1,1))


```

```{r}
SE_male_dupRM_25up$response <- factor(GMM_male$classification[match(SE_male_dupRM_25up$ID, names(GMM_male$classification))],  levels = c(1,2))
```

#### On female cohort

Code are not shown in html to reduce redundancy, please refer to Rmd file for code.

```{r, echo = FALSE, message = FALSE}
female_exprs_day1_0 <- list( day1 = SE_female_dupRM_25up[which(rownames(SE_female_dupRM_25up) %in% DE11_response),which(SE_female_dupRM_25up$time == "Day1")],
                           day0 = SE_female_dupRM_25up[which(rownames(SE_female_dupRM_25up) %in% DE11_response),which(SE_female_dupRM_25up$time == "Day0")])
colnames(female_exprs_day1_0[["day1"]]) <- female_exprs_day1_0[["day1"]]$ID
colnames(female_exprs_day1_0[["day0"]]) <- female_exprs_day1_0[["day0"]]$ID

female_exprs_day1_0[["day1"]] <- as.matrix(assay(female_exprs_day1_0[["day1"]]))
female_exprs_day1_0[["day0"]] <- as.matrix(assay(female_exprs_day1_0[["day0"]]))

# Match the order of gene and order of participants 
female_exprs_day1_0[[2]] <- female_exprs_day1_0[[2]][match(rownames(female_exprs_day1_0[[1]]), rownames(female_exprs_day1_0[[2]])), match(colnames(female_exprs_day1_0[[1]]), colnames(female_exprs_day1_0[[2]]))]

# Subtract day 0 from day 1 for each participants
female_diff_day1_0 <- (female_exprs_day1_0[["day1"]] - female_exprs_day1_0[["day0"]])
```

GMM result.

```{r, echo = FALSE}
# GMM cluster
GMM_female <-  Mclust(t(female_diff_day1_0),  G = 2)
table(GMM_female$classification)
```

Overall, we assign group 2 as high responders and group 1 as low responders for female cohort based on boxplots. 

```{r, fig.width= 8, fig.height=5, echo = FALSE}

# Check GMM cluster result wrt 10 genes
par(mfrow=c(2,5))
for (i in 1:length(names_table1)){
  dff <- data.frame(expr = female_diff_day1_0[names_table1[i],], group=GMM_female$classification)
  boxplot(expr ~group,dff, main=names_table1[i], ylab="diff(day1- day0)")
}
par(mfrow=c(1,1))


```

```{r, echo = FALSE}
SE_female_dupRM_25up$response <- factor(GMM_female$classification[match(SE_female_dupRM_25up$ID, names(GMM_female$classification))],  levels = c(1,2))
```

# Results

We first define few functions used for analysis.

```{r}

mixed_test <- function(gene,gene_name, covariates, fml, var_rowname, j){
  df_test <- cbind(gene , covariates)
  colnames(df_test)[1] <- "gene"
  mix_model <- aov(fml, data=df_test)
  df_ret <- data.frame(unlist(summary(mix_model))[paste0("Error: Within.Pr(>F)",j)])
  rownames(df_ret) <- var_rowname
  colnames(df_ret) <- gene_name
  return(tibble(df_ret))
}


GO_analysis <- function(genes){
  DE <- gost(query = genes, organism = "hsapiens", multi_query = FALSE,
             correction_method = "fdr", user_threshold = 0.01,  
             significant = TRUE,
             source = "GO:BP")

  print(table(DE[["result"]][["source"]]))

  DE_mod = DE$result[,c("query",  "term_id",
                                "term_name", "p_value", "query_size", 
                                "intersection_size", "term_size", 
                                "effective_domain_size", "intersection_size", "source")]

  DE_mod$"GeneRatio" = paste0(DE_mod$intersection_size,  "/", DE_mod$query_size)
  DE_mod$BgRatio = paste0(DE_mod$term_size, "/", DE_mod$effective_domain_size)
  DE_mod$"pvalue" <- DE_mod$p_value
  DE_mod$Count <- DE_mod$intersection_size
  DE_mod$Description<- DE_mod$term_name
  DE_mod_enrich <-  new("enrichResult", result = DE_mod)

  dotplot(DE_mod_enrich , x = 'GeneRatio', color = "pvalue") + ggtitle("GO: Biological process")
}

```




## Male cohort

### Analysis with respect to time

We first fit mixed effect model, then extract p-value from testing the fix effect on time which is treated as categorical variable. 


```{r}

cov_male <- data.frame(ID = SE_male_dupRM_25up$ID,
                  time = SE_male_dupRM_25up$time,
                  response = SE_male_dupRM_25up$response)

genes_male <- rownames(SE_male_dupRM_25up)
mix_test_time_male <- sapply( 1: nrow(SE_male_dupRM_25up),
                            function(i) mixed_test( gene = as.vector(SE_male_dupRM_25up@assays@data@listData[["exprs"]][i,]),
                                                         gene_name = genes_male[i],
                                                         covariates = cov_male,
                                                         fml = as.formula(gene ~ time + Error(ID)),
                                                         var_rowname = c("time"),
                                                         j = 1))
mix_test_time_male <- as.data.frame(unlist(mix_test_time_male))  
colnames(mix_test_time_male) <- c("pvalue")

hist(mix_test_time_male$pvalue, main ="P-value from fix effect of time", xlab="P-value")

```

P-values are adjusted using Benjamini-Hotchberg (BH) method.

```{r}
# BH adjusted p-value
mix_test_time_male$BH_pvalue <- p.adjust(mix_test_time_male$pvalue, method =  "BH", n = length(mix_test_time_male$pvalue))

hist(mix_test_time_male$BH_pvalue, main ="BH adjusted p-value from fix effect of time", xlab="BH adjusted p-value")

```

Controlling for 1% of false discovery rate, there are 4540 differentially expressed (DE) genes across time in the male cohort.

```{r}
DEtime_male <- mix_test_time_male %>% as.data.frame() %>% filter(BH_pvalue < 0.01)

nrow(DEtime_male)
```

Next, we plot heatmap on top 50 DE genes by BH adjusted p-value. For male cohort, the top 50 DE genes are predominantly up-regulated at day 1 after vaccination. This is consistent with what authors reported in their paper. 

```{r, fig.height=10, fig.width=8}
DE50_male <- SE_male_dupRM_25up[which(rownames(SE_male_dupRM_25up) %in% rownames(DEtime_male %>% arrange(BH_pvalue))[1:50]), ]

colData(DE50_male)$time <- factor(colData(DE50_male)$time, levels = c("Day0", "Day1", "Day3", "Day14"))

DE50_male_mean <- sapply(c("Day0", "Day1", "Day3", "Day14"), function(day) rowMeans(assay(DE50_male)[, which(colData(DE50_male)$time == day)]))

pheatmap(DE50_male_mean, cluster_rows = TRUE,  cluster_cols = FALSE, scale = "row", 
          main = "Mean expression of top 50 DE genes wrt time by BH adjusted p-values")

```

Next, we perform Gene Ontology enrichment analysis on the 4540 DE genes. Specifically, we are interested in biological process. 

There are 1550 GO terms showed significant association with 4540 DE genes found with respect to time. Some of the top GO term related to immune response and stress are what we expected after vaccination. 

```{r}

GO_analysis(rownames(DEtime_male))
```

### Analysis with respect to response across time

We fit mixed effect model described above, then extract p-values and adjust by using Benjamini-Hochberg method.


```{r}

mix_test_response_time_male <- sapply( 1: nrow(SE_male_dupRM_25up),
                            function(i) mixed_test( gene = as.vector(SE_male_dupRM_25up@assays@data@listData[["exprs"]][i,]),
                                                         gene_name = genes_male[i],
                                                         covariates = cov_male,
                                                         fml = as.formula(gene ~ time*response + Error(ID)),
                                                         var_rowname = c("time"),
                                                         j = 2))

mix_test_response_time_male <- as.data.frame(unlist(mix_test_response_time_male))

colnames(mix_test_response_time_male) <- c("pvalue")

hist(mix_test_response_time_male$pvalue, main ="P-value from fix effect of time*response", xlab="P-value")

# BH adjusted p-value
mix_test_response_time_male$BH_pvalue <- p.adjust(mix_test_response_time_male$pvalue, method =  "BH", n = length(mix_test_response_time_male$pvalue))

hist(mix_test_response_time_male$BH_pvalue, main ="BH adjusted p-value from fix effect of time*response", xlab="BH adjusted p-value")

```

Controlling for 5% of false discovery rate, there are 573 differentially expressed (DE) genes across time in the male cohort.

```{r}
DEresponse_male <- mix_test_response_time_male %>% as.data.frame() %>% filter(BH_pvalue < 0.05)

nrow(DEresponse_male)
```

We plot heatmap on top 50 DE genes by BH adjusted p-value. For male cohort, the top 50 DE genes are predominantly up-regulated at day 1 among high responders after vaccination. There are few genes that are down-regulated at day 1 among high responders after vaccination.

```{r, fig.height=10, fig.width=8}
DE50response_male <- SE_male_dupRM_25up[which(rownames(SE_male_dupRM_25up) %in% rownames(DEresponse_male %>% arrange(BH_pvalue))[1:50]), ]

colData(DE50response_male)$time <- factor(colData(DE50response_male)$time, levels = c("Day0", "Day1", "Day3", "Day14"))

DE_response1_mean_male <- sapply(c( "Day0", "Day1", "Day3", "Day14"), function(day) rowMeans(assay(
DE50response_male)[, which(colData(
DE50response_male)$time == day & DE50response_male$response == 1)]))
colnames(DE_response1_mean_male) <- paste0(colnames(DE_response1_mean_male), "_low")

DE_response2_mean_male <- sapply(c( "Day0", "Day1", "Day3", "Day14"), function(day) rowMeans(assay(
DE50response_male)[, which(colData(
DE50response_male)$time == day & DE50response_male$response == 2)]))
colnames(DE_response2_mean_male) <- paste0(colnames(DE_response2_mean_male), "_high")

Group_male <- data.frame( Group = c(rep("Low responder",4), rep("High responder", 4)))
rownames(Group_male) <- c(colnames(DE_response1_mean_male), colnames(DE_response2_mean_male))

pheatmap(cbind(DE_response1_mean_male,DE_response2_mean_male), 
         cluster_rows = TRUE,  cluster_cols = FALSE, scale = "row", 
         main = "Mean expression on top 50 DE genes wrt response over time", 
         annotation_col = Group_male)


```

From GO enrichment analysis, there are 483 GO terms showed significant association with 573 DE genes found with respect to response over time. The top GO terms are mostly related to immune response.

```{r}
GO_analysis(rownames(DEresponse_male))

```

## Female cohort

Code for analysis on female cohort are not shown in html, please refer to Rmd file for code.

### Analysis with respect to time

We fit mixed effect model, and extract p-value on testing the fixed effect of interaction between response and time. P-values are further adjusted using Benjamini-Hotchberg method.

```{r, echo = FALSE}

cov_female <- data.frame(ID = SE_female_dupRM_25up$ID,
                  time = SE_female_dupRM_25up$time,
                  response = SE_female_dupRM_25up$response)

genes_female <- rownames(SE_female_dupRM_25up)
mix_test_time_female <- sapply( 1: nrow(SE_female_dupRM_25up),
                            function(i) mixed_test( gene = as.vector(SE_female_dupRM_25up@assays@data@listData[["exprs"]][i,]),
                                                         gene_name = genes_female[i],
                                                         covariates = cov_female,
                                                         fml = as.formula(gene ~ time + Error(ID)),
                                                         var_rowname = c("time"),
                                                         j = 1))
mix_test_time_female <- as.data.frame(unlist(mix_test_time_female))  
colnames(mix_test_time_female) <- c("pvalue")

hist(mix_test_time_female$pvalue, main ="P-value from fix effect of time", xlab="P-value")

```

```{r, echo = FALSE}
# BH adjusted p-value
mix_test_time_female$BH_pvalue <- p.adjust(mix_test_time_female$pvalue, method =  "BH", n = length(mix_test_time_female$pvalue))

hist(mix_test_time_female$BH_pvalue, main ="BH adjusted p-value from fix effect of time", xlab="BH adjusted p-value")

```

Controlling for 1% of false discovery rate, there are 6118 differentially expressed (DE) genes across time in the male cohort.

```{r , echo = FALSE}
DEtime_female <- mix_test_time_female %>% as.data.frame() %>% filter(BH_pvalue < 0.01)

nrow(DEtime_female)
```

We plot heatmap on top 50 DE genes by BH adjusted p-value. For female cohort, the top 50 DE genes are all up-regulated at day 1 after vaccination in female cohort. 

```{r, fig.height=10, fig.width=8 , echo = FALSE}
DE50_female <- SE_female_dupRM_25up[which(rownames(SE_female_dupRM_25up) %in% rownames(DEtime_female %>% arrange(BH_pvalue))[1:50]), ]

colData(DE50_female)$time <- factor(colData(DE50_female)$time, levels = c("Day0", "Day1", "Day3", "Day14"))

DE50_male_femean <- sapply(c("Day0", "Day1", "Day3", "Day14"), function(day) rowMeans(assay(DE50_female)[, which(colData(DE50_female)$time == day)]))

pheatmap(DE50_male_femean, cluster_rows = TRUE,  cluster_cols = FALSE, scale = "row", 
          main = "Mean expression of top 50 DE genes wrt time by BH adjusted p-values")

```

There are 1682 GO terms showed significant association with 6118 DE genes found with respect to time. Some of the top GO term related to immune response and stress are what we expected after vaccination. 

```{r , echo = FALSE}

GO_analysis(rownames(DEtime_female))
```

### Analysis with respect to response across time

Response to vaccination is inferred using GMM in previous section. We fit a mixed effect model including interaction between response and time as fixed effect. Then p-values on testing the effect of the interaction are extracted, then adjusted using Benjamini-Hotchberg method.

```{r, echo = FALSE}

mix_test_response_time_female <- sapply( 1: nrow(SE_female_dupRM_25up),
                            function(i) mixed_test( gene = as.vector(SE_female_dupRM_25up@assays@data@listData[["exprs"]][i,]),
                                                         gene_name = genes_female[i],
                                                         covariates = cov_female,
                                                         fml = as.formula(gene ~ time*response + Error(ID)),
                                                         var_rowname = c("time"),
                                                         j = 2))

mix_test_response_time_female <- as.data.frame(unlist(mix_test_response_time_female))

colnames(mix_test_response_time_female) <- c("pvalue")

hist(mix_test_response_time_female$pvalue, main ="P-value from fix effect of time*response", xlab="P-value")

# BH adjusted p-value
mix_test_response_time_female$BH_pvalue <- p.adjust(mix_test_response_time_female$pvalue, method =  "BH", n = length(mix_test_response_time_female$pvalue))

hist(mix_test_response_time_female$BH_pvalue, main ="BH adjusted p-value from fix effect of time*response", xlab="BH adjusted p-value")

```


Controlling for 5% of false discovery rate, there are 3217 differentially expressed (DE) genes with respect to response across time in the female cohort.

```{r, echo = FALSE}
DEresponse_female <- mix_test_response_time_female %>% as.data.frame() %>% filter(BH_pvalue < 0.05)

nrow(DEresponse_female)
```

From the heatmap on top 50 DE genes by BH adjusted p-value, majority of the DE genes are down-regulated at all time points among high responders after vaccination. This is different than the result shown for male analysis.

```{r, fig.height=10, fig.width=8, echo = FALSE}
DE50response_female <- SE_female_dupRM_25up[which(rownames(SE_female_dupRM_25up) %in% rownames(DEresponse_female %>% arrange(BH_pvalue))[1:50]), ]

colData(DE50response_female)$time <- factor(colData(DE50response_female)$time, levels = c("Day0", "Day1", "Day3", "Day14"))

DE_response1_mean_female <- sapply(c( "Day0", "Day1", "Day3", "Day14"), function(day) rowMeans(assay(
DE50response_female)[, which(colData(
DE50response_female)$time == day & DE50response_female$response == 1)]))
colnames(DE_response1_mean_female) <- paste0(colnames(DE_response1_mean_female), "_low")

DE_response2_mean_female <- sapply(c( "Day0", "Day1", "Day3", "Day14"), function(day) rowMeans(assay(
DE50response_female)[, which(colData(
DE50response_female)$time == day & DE50response_female$response == 2)]))
colnames(DE_response2_mean_female) <- paste0(colnames(DE_response2_mean_female), "_high")

Group_female <- data.frame( Group = c(rep("Low responder",4), rep("High responder", 4)))
rownames(Group_female) <- c(colnames(DE_response1_mean_female), colnames(DE_response2_mean_female))

pheatmap(cbind(DE_response1_mean_female,DE_response2_mean_female), 
         cluster_rows = TRUE,  cluster_cols = FALSE, scale = "row", 
         main = "Mean expression on top 50 DE genes wrt response over time", 
         annotation_col = Group_female)


```

There are 730 GO terms showed significant association with 3217 DE genes found with respect to response over time. The top GO terms are different as compared to the analysis on male cohort. 

```{r, echo = FALSE}
GO_analysis(rownames(DEresponse_female))

```


# Conclusion and discussion

The overarching results expressed in the original paper were replicated well for the male cohort, despite having incomplete information on actual response status. Differential expression peaked on Day 1 post vaccination, and the majority of genes that were differentially expressed when accounting for response status were related to immunological biological processes. This result was not consistent in the female cohort, where subjects assigned a high response showed over-expression at Day 0 compared to post-vaccination, and none of the top Gene Ontology terms related to biological process involved the immune response. This may indicate a difference in sexes, but is more likely an artifact of misclassification errors, as the 11 genes used to predict response category were identified in the male data. The response patterns over time, without respect to response status, look similar in terms of expression changes, significance, and GO terms, between males and females.

The two main takeaways from this data are the overall timing of the expression patterns, and potential pathways to examine to see if they can be targeted in the development of novel vaccines or adjuvants (molecules co-administered with vaccines to enhance immune response). In particular, the results of this study indicate that sampling for gene expression at Day 1 post vaccination could be a useful addition in early-phase vaccine studies, as expression at this timepoint gives the most information about the pathways triggered, and could in extreme cases (such as future pandemics) provide immediate insight into vaccine efficacy, should expression correlates be well established.

# Reference

Bucasas, K. L., Franco, L. M., Shaw, C. A., Bray, M. S., Wells, J. M., Niño, D., Arden, N., Quarles, J. M., Couch, R. B., &amp; Belmont, J. W. (2011). Early patterns of gene expression correlate with the humoral immune response to influenza vaccination in humans. The Journal of Infectious Diseases, 203(7), 921–929. https://doi.org/10.1093/infdis/jiq156  