---
title: "Gene Expression Response to Influenza Vaccination"
author: "Kaitlyn Cross, Eunchong Kang, Yu (Sylvia) Zhang"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
  wrap: 80
---

```{r load library, message = FALSE}
library(ggplot2)
library(dplyr)
library(rmarkdown)
library(knitr)
library(GEOquery)
library(SummarizedExperiment)
library(DESeq2)
library(here)
library(limma)
library(nlme)
library(MASS)
library(multcomp)
library(pheatmap)
library(variancePartition)
library(lme4)
library("r2glmm")
```

# Introduction

# Data collection

Data for the study on trivalent influenza vaccine are deposited in National Center for Biotechnology Information (NCBI) Gene Expression Omnibus (GEO) with accession GSE48024. A total of 119 male and 128 female between 19 to 41 years of age were recruited in the study. Whole blood samples were drawn before vaccination (day 0) and at three time points after vaccination (day 1, 3 and 14). All male samples are sequenced using Illumina HumanHT-12 V3.0 expression beadchip and all female samples are sequence using Illumina HumanHT-12 V4.0 expression beadchip. 

## Load data

Two separate datasets are loaded, one for female cohort and another one for male cohort.

```{r}
gset <- getGEO("GSE48024", GSEMatrix =TRUE, AnnotGPL=TRUE)

gset_female <- gset[[1]]
gset_male <- gset[[2]]

# Check gender 
unique(gset_female$`gender:ch1`) 
unique(gset_male$`gender:ch1`)

# Check sequencing platform 
unique(gset_female$platform_id) #Illumina HumanHT-12 V4.0 expression beadchip
unique(gset_male$platform_id) #Illumina HumanHT-12 V3.0 expression beadchip

```

Based on the data loaded directly from GEO, 116 out of 119 males and 110 out of 128 females are available. A total of 417 samples from female cohort and 431 samples from male cohort.

```{r}
# Check number of subjects available in data
length(unique(gset_female$`subject:ch1`))
length(unique(gset_male$`subject:ch1`))

# Check number of samples available in data
ncol(gset_female)
ncol(gset_male)
```

There are 47276 probes used for female cohort, and 48742 probes used for male cohort.

```{r}
nrow(gset_female)
nrow(gset_male)
```

Data has been processed and normalized according to desciprtion given in GEO. We randomly select 10 samples from each cohort to check the distribution of normalized data.

```{r}
# Data processing + normalization 
unique(gset_male$data_processing)
unique(gset_female$data_processing)

boxplot(exprs(gset_male)[,sample(1:431, 10, replace=F)], 
        main = "Ten random samples from male cohort", ylab = "Gene expression")

boxplot(exprs(gset_female)[,sample(1:417, 10, replace=F)], 
        main = "Ten random samples from female cohort", ylab = "Gene expression")

```

Last, we check how many samples at each sampling time for each cohort.

```{r}
# For female
table(gset_female$`time:ch1`)

# For male
table(gset_male$`time:ch1`)

```

# Analysis plan

# Pre-processing

# Results

## DE

## GO analysis

# Conclusion and discussion



