---
title: "DE_analysis_male"
author: "Yu (Sylvia) Zhang"
date: "`r Sys.Date()`"
output: 
  html_document:
        toc: true
        toc_float: true
editor_options: 
  chunk_output_type: console
---

```{r load package, message = FALSE}
library(GEOquery)
library(SummarizedExperiment)
library(DESeq2)
library(here)
library(dplyr)
library(limma)
library(nlme)
library(MASS)
library(multcomp)
library(pheatmap)
library(variancePartition)
library(lme4)
library("r2glmm")
```

# Load data

I saved the data into a local directory to save time from downloading each time.

```{r load data, eval = FALSE}
gset <- getGEO("GSE48024", GSEMatrix =TRUE, AnnotGPL=TRUE)
save(gset, file = "/Users/yu.zhang/Desktop/BIOS 784/Data/GSE48024/gset.RData")
```

```{r}
load("/Users/yu.zhang/Desktop/BIOS 784/Data/GSE48024/gset.RData")
```

### Check data

```{r, eval = FALSE}
gset_male <- gset[[2]]

length(unique(gset_male$`subject:ch1`))
table(gset_male$`time:ch1`)

male_all4 <- data.frame(table(gset_male$`subject:ch1`)) %>% filter(Freq == 4)

gset_female <- gset[[1]]

length(unique(gset_female$`subject:ch1`))
table(gset_female$`time:ch1`)

female_all4 <- data.frame(table(gset_female$`subject:ch1`)) %>% filter(Freq == 4)
```

# Working on male data

```{r}
gset_male <- gset[[2]]
```

```{r}
boxplot(exprs(gset_male)[,sample(1:431, 10, replace=F)], 
        main = "Ten random samples from male cohort")
```

## Subset to probes with gene name

Total 431 samples and 29228 probes with valid gene symbol.

```{r}
gset_male <- gset_male[which(gset_male@featureData@data[["Gene symbol"]] != ""), ]

gset_male
```

## Set up summarized experiment object

```{r}
SE_male  <- SummarizedExperiment(assay = list("exprs" = exprs(gset_male)),
                                  colData = as(gset_male@phenoData, "data.frame"),
                                  rowData = as(gset_male@featureData, "data.frame"))

table(SE_male$`time:ch1`)
```

## Subset to participants sampled at all 4 time points

```{r}
paticipants.all <- as.data.frame(table(SE_male$`subject:ch1`)) %>% filter(Freq == 4)

SE_male <- SE_male[, which(SE_male$`subject:ch1` %in% paticipants.all$Var1)]

table(SE_male$`time:ch1`)

```

## Merge duplicate probes, averaging expression from the same gene

Said from limma for averaging expression: "This function should only be applied to normalized log-expression values, and not to raw unlogged expression values. It will generate an error message if applied to RGList or EListRaw objects."

```{r}
rownames(SE_male)[1:10]

expression.dupRM <- avereps(SE_male@assays@data@listData[["exprs"]], ID=SE_male@elementMetadata@listData[["Gene symbol"]])

# match(colnames(expression.dupRM ) , rownames(SE_male@colData))

SE_male_dupRM <- SummarizedExperiment(assay = list("exprs" = expression.dupRM),
                                  colData = SE_male@colData)

SE_male_dupRM
```

## Remove lower 25% variant genes

```{r}
SE_male_dupRM@metadata[["gene_mean"]] <- rowMeans(SE_male_dupRM@assays@data@listData[["exprs"]])

boxplot(SE_male_dupRM@metadata[["gene_mean"]])

SE_male_dupRM@metadata[["gene_variance"]] <- rowVars(SE_male_dupRM@assays@data@listData[["exprs"]])

boxplot(SE_male_dupRM@metadata[["gene_variance"]])

quantile(SE_male_dupRM@metadata[["gene_mean"]]) # 25% = 7.734803
quantile(SE_male_dupRM@metadata[["gene_variance"]]) # 25% =  0.0022386139, 75% = 0.0134488159

SE_male_dupRM_25up <- SE_male_dupRM[which(SE_male_dupRM@metadata[["gene_variance"]] > 0.0022386139), ]
SE_male_dupRM_25up

#SE_male_dupRM_75up <- SE_male_dupRM[which(SE_male_dupRM@metadata[["gene_variance"]] > 0.0134488159), ]
#SE_male_dupRM_75up

#save(SE_male_dupRM_25up, file = "/Users/yu.zhang/Desktop/BIOS 784/Data/GSE48024/SE_male_dupRM_25up.RData")
#save(SE_male_dupRM_75up, file = "/Users/yu.zhang/Desktop/BIOS 784/Data/GSE48024/SE_male_dupRM_75up.RData")
```

After filtering out lower 25%, we are left with 14692 unique genes and 368 samples.

After filtering out lower 50%, we are left with 9795 unique genes and 368 samples.

After filtering out lower 75%, we are left with 4898 unique genes and 368 samples.

# DE analysis wrt time only

## Function for DE analysis using top 25% most variant genes


```{r}
mixed_test <- function(gene,gene_name, covariates, fml, var_rowname){

  df_test <- cbind(gene , covariates)
  colnames(df_test)[1] <- "gene"
  mix_model <- aov(fml, data=df_test)
  df_ret <- data.frame(unlist(summary(mix_model))["Error: Within.Pr(>F)1"])
  rownames(df_ret) <- var_rowname
  colnames(df_ret) <- gene_name
  return(tibble(df_ret))
}

mixed_test2 <- function(gene,gene_name, covariates, fml, var_rowname){
  
   #gene = as.data.frame(SE_male_dupRM_25up@assays@data@listData[["exprs"]][which(rownames(SE_male_dupRM_25up) ==  genes[1]),])
  #gene_name = genes[1]
  #covariates = cov
  #fml = as.formula(gene ~ time*response + Error(ID))
  #var_rowname = c("response_time")
  
  df_test <- cbind(gene , covariates)
  colnames(df_test)[1] <- "gene"
  mix_model <- aov(fml, data=df_test)
  df_ret <- data.frame(unlist(summary(mix_model))["Error: Within.Pr(>F)2"])
  rownames(df_ret) <- var_rowname
  colnames(df_ret) <- gene_name
  return(tibble(df_ret))
}

```

## Mixed effect model, gene ~ time + 1|ID

#### P-value from mixed effect model

```{r, eval = FALSE}
genes <- rownames(SE_male_dupRM_25up)

cov <- data.frame(ID = SE_male_dupRM_25up$`subject:ch1`,
                  gender = SE_male_dupRM_25up$`gender:ch1`,
                  time = as.character(SE_male_dupRM_25up$`time:ch1`))


mix_test_time <- sapply( 1:nrow(SE_male_dupRM_25up), function(i) mixed_test(   gene = as.data.frame(SE_male_dupRM_25up@assays@data@listData[["exprs"]][which(rownames(SE_male_dupRM_25up) ==  genes[i]),]),
  gene_name = genes[i],
  covariates = cov,
  fml = as.formula(gene ~ time + Error(ID)),
  var_rowname = c("time")) )

mix_test_time <- as.data.frame(unlist(mix_test_time))

colnames(mix_test_time) <- c("p_value_mix_compsym")

write.csv(mix_test_time, file = paste0(here(), "/code/2_DE_analysis/mix_pvalue_compsym.csv"))


```

#### BH-adjusted p-value

```{r}
mix_test_time <- read.csv(file = paste0(here(), "/code/2_DE_analysis/mix_pvalue_compsym.csv"))
rownames(mix_test_time) <- mix_test_time$X

boxplot(mix_test_time$p_value_mix_compsym, main ="P-value from mixed model, gene ~ time + (1|ID)")

hist(mix_test_time$p_value_mix_compsym, main ="P-value from mixed model, gene ~ time + (1|ID)")

# BH adjusted p-value
mix_test_time$BH_pvalue <- p.adjust(mix_test_time$p_value_mix_compsym, method =  "BH", n = length(mix_test_time$p_value_mix_compsym))

boxplot(mix_test_time$BH_pvalue, main ="BH adjusted P-value from mixed model, gene ~ time + 1|ID")

hist(mix_test_time$BH_pvalue, main ="BH adjusted P-value from mixed model, gene ~ time + 1|ID")

DEgene_pvalue_BH <- mix_test_time %>% as.data.frame() %>% filter(BH_pvalue < 0.01)

nrow(DEgene_pvalue_BH)

```
A total of 4540 DE genes wrt time after BH adjustment.

#### Heatmap on top 50 DE genes from BH adjusted p-values

```{r, fig.height=8, fig.width=6 }

DE_genes <- SE_male_dupRM_25up[which(rownames(SE_male_dupRM_25up) %in% rownames(DEgene_pvalue_BH %>% arrange(BH_pvalue))[1:50]), ]

colData(DE_genes)$time <- factor(colData(DE_genes)$`time:ch1`, levels = c("Day0", "Day1", "Day3", "Day14"))

table(colData(DE_genes)$time)

DE_genes_mean <- sapply(c("Day0", "Day1", "Day3", "Day14"), function(day) rowMeans(assay(DE_genes)[, which(colData(DE_genes)$time == day)]))

pheatmap(DE_genes_mean, cluster_rows = TRUE,  cluster_cols = FALSE, scale = "row", 
          main = "Heatmap on top 50 DE genes wrt time by BH adjusted p-values")

```

### Gene ontology enrichment analysis

```{r, message = FALSE}
library(enrichplot)
library(DOSE)
library(gprofiler2)

DE_gene_name <- gost(query = DEgene_pvalue_BH$X, organism = "hsapiens", multi_query = FALSE,
                     correction_method = "fdr", user_threshold = 0.01, source = "GO:BP")

head(DE_gene_name$result)

table(DE_gene_name[["result"]][["source"]])

#gostplot(DE_gene_name, interactive = TRUE)

colnames(DE_gene_name$result)

gostplot(DE_gene_name, capped = TRUE, interactive = FALSE)

DE_mod = DE_gene_name$result[,c("query",  "term_id",
                                "term_name", "p_value", "query_size", 
                                "intersection_size", "term_size", 
                                "effective_domain_size", "intersection_size", "source")]

DE_mod$"GeneRatio" = paste0(DE_mod$intersection_size,  "/", DE_mod$query_size)
DE_mod$BgRatio = paste0(DE_mod$term_size, "/", DE_mod$effective_domain_size)
DE_mod$"pvalue" <- DE_mod$p_value
DE_mod$Count <- DE_mod$intersection_size
DE_mod$Description<- DE_mod$term_name
DE_mod_enrich  <-  new("enrichResult", result = DE_mod)

dotplot(DE_mod_enrich , x = 'GeneRatio', color = "pvalue") + ggtitle("GO: Biological process")

#BP <-  new("enrichResult", result = DE_mod_enrich@result[which(DE_mod_enrich@result[["source"]] == "GO:BP") , ])

#dotplot(BP , x = 'GeneRatio', color = "pvalue")

```

```{r test code, echo=FALSE,eval = FALSE}
library(topGO)
library("AnnotationHub")
library(GO.db)
library(clusterProfiler)

data(geneList)
geneList[1:5]
de <- names(geneList)[1:100]
de[1:5]
x <- enrichDO(de)
dotplot(x)

# Maybe useful? 
ah <- AnnotationHub()
orgs <- subset(ah, ah$rdataclass == "OrgDb")
orgdb <- query(orgs, "Homo sapiens")[[1]]
geneList <- DEgene_pvalue_BH$BH_pvalue
names(geneList) <- DEgene_pvalue_BH$X
GOdata <- new("topGOdata",
    ontology = "BP",
    allGenes = geneList,
    geneSelectionFun = function(x)x,
    annot = annFUN.org, mapping = )

```

# Cluster participants based on diff(day1 - day0) expression 

```{r }

library(mclust)
library(factoextra)

table(SE_male_dupRM_25up$`time:ch1`)

DE10_response <- c("STAT1", "IRF9", "SPI1", "CD74", "HLA-E", "TNFSF13B", "PRDX2", "PRDX3", "E2F2", "PTEN", "ITGB1")

gene_exprs_day1_0 <- list( day1 = SE_male_dupRM_25up[,which(SE_male_dupRM_25up$`time:ch1` == "Day1")],
                           day0 = SE_male_dupRM_25up[,which(SE_male_dupRM_25up$`time:ch1` == "Day0")])

dim(gene_exprs_day1_0[[1]])

gene_exprs_day1_0[[2]] <- gene_exprs_day1_0[[2]][match(rownames(gene_exprs_day1_0[[2]]), rownames(gene_exprs_day1_0[[1]])), match(gene_exprs_day1_0[[2]]$`subject:ch1`, gene_exprs_day1_0[[1]]$`subject:ch1`)]

diff_day1_0 <- gene_exprs_day1_0[["day1"]]@assays@data@listData[["exprs"]] - gene_exprs_day1_0[["day0"]]@assays@data@listData[["exprs"]]

diff_day1_0 <- diff_day1_0[which(rownames(diff_day1_0) %in% DE10_response) , ]

#hist(diff_day1_0[1,], breaks=30)

colnames(diff_day1_0) <- gene_exprs_day1_0[[2]]$`subject:ch1`

GMM <-  Mclust(t(diff_day1_0),  G = 2)
table(GMM$classification)

km<- kmeans(t(diff_day1_0), 2, nstart = 25)
table(km$cluster)

table(GMM$classification , km$cluster)

SE_male_dupRM_25up$response <- GMM$classification[match(SE_male_dupRM_25up$`subject:ch1`, names(GMM$classification)) ]
```

Out of 92 participants, GMM and K-means agree on the classification of 24+49=73 participants.  

For downstream analysis of DE genes on response over time, I'm using result from GMM for now. If we are interested in K-means after I put more visualization, I can definitely switch. 

# possible relationship between clustering results and responsiveness
```{r }
# Gene names in Table 1
names_table1 <- c("STAT1", "IRF9", "SPI1", "CD74", "HLA-E", "TNFSF13B", "PRDX3", "E2F2", "PTEN", "ITGB1")

# Boxplot of diff(day1-day0) by clustered group across the genes in Table 1
par(mfrow=c(2,5))
for (i in 1:length(names_table1)){
  dff <- data.frame(expr = diff_day1_0[names_table1[i],], group=GMM$classification)
  boxplot(expr ~group,dff, main=names_table1[i], ylab="diff(day1- day0)")
}
par(mfrow=c(1,1))

# Create Table1
table1 <- data.frame(Gene=names_table1, "Upregulated in"=c("High responders", "High responders","High responders","High responders","High responders","High responders","Low responders","Low responders","Low responders","Low responders"))

## Subset of Transcripts Used for prediction of the TRI
table1
```

Interestingly, the results correspond to Table 1 assuming response 2 means the high response. For example, the paper says STAT1 gene is upregulated in the high responders, and the boxplot shows the diff(day1-day0) for STAT1 is higher in the response 2 (high responder) with the value greater than 0 than one in response 1 (low responder) with the value of around 0. The only exception is PTEN, which is said to be upregulated in the low responders but the boxplot shows the diff is higher and greater than 0 in response 2 (high responder). Although there is no solid evidence that the clustering result represents true responsiveness, we have some coincident results between the paper and our clustering and help us move on to the next analysis assuming that clustering is partly based on responsiveness.

# DE genes on response over time, using Day 1 - 14 data only

```{r, eval = FALSE}

genes <- rownames(SE_male_dupRM_25up)

cov <- data.frame(ID = SE_male_dupRM_25up$`subject:ch1`,
                  gender = SE_male_dupRM_25up$`gender:ch1`,
                  time = factor(SE_male_dupRM_25up$`time:ch1`, level = c("Day0", "Day1", "Day3", "Day14")),
                  response = factor(SE_male_dupRM_25up$response, level = c(1,2)))

mix_test_response_time <- sapply( 1:nrow(SE_male_dupRM_25up), function(i) mixed_test2( gene = as.data.frame(SE_male_dupRM_25up@assays@data@listData[["exprs"]][which(rownames(SE_male_dupRM_25up) ==  genes[i]),]),
  gene_name = genes[i],
  covariates = cov,
  fml = as.formula(gene ~ time*response + Error(ID)),
  var_rowname = c("response_time")) )

mix_test_response_time <- as.data.frame(unlist(mix_test_response_time))

colnames(mix_test_response_time) <- c("p_value_mix_compsym")

write.csv(mix_test_response_time, file = paste0(here(), "/code/2_DE_analysis/mix_response_time_pvalue_compsym.csv"))

```


#### BH-adjusted p-value

57 DE genes on response over time for BH p-value cutoff = 0.05.

340 DE genes on response over time for BH p-value cutoff = 0.1.

```{r}
mix_test_response_time <- read.csv(file = paste0(here(), "/code/2_DE_analysis/mix_response_time_pvalue_compsym.csv"))
rownames(mix_test_response_time) <- mix_test_response_time$X

boxplot(mix_test_response_time$p_value_mix_compsym, main ="P-value from mixed model, gene ~ time*response + (1|ID)")

hist(mix_test_response_time$p_value_mix_compsym, main ="P-value from mixed model, gene ~ time*response + (1|ID)")

# BH adjusted p-value
mix_test_response_time$BH_pvalue <- p.adjust(mix_test_response_time$p_value_mix_compsym, method =  "BH", n = length(mix_test_response_time$p_value_mix_compsym))

boxplot(mix_test_response_time$BH_pvalue, main ="BH adjusted P-value from mixed model, gene ~ time*response + 1|ID")

hist(mix_test_response_time$BH_pvalue, main ="BH adjusted P-value from mixed model, gene ~ time*response + 1|ID")

DEgene_pvalue_BH_response <- mix_test_response_time %>% as.data.frame() %>% filter(BH_pvalue < 0.05)

nrow(DEgene_pvalue_BH_response)

```


#### Heatmap, this is interesting


```{r, fig.height=8, fig.width=8}

DE_genes_response <- SE_male_dupRM_25up[which(rownames(SE_male_dupRM_25up) %in% rownames(DEgene_pvalue_BH_response%>% arrange(BH_pvalue))[1:50] ), ]

colData(DE_genes_response)$time <- factor(colData(DE_genes_response)$`time:ch1`, levels = c("Day0", "Day1", "Day3", "Day14"))

table(colData(DE_genes_response)$time)

DE_genes_response1_mean <- sapply(c( "Day0", "Day1", "Day3", "Day14"), function(day) rowMeans(assay(
DE_genes_response)[, which(colData(
DE_genes_response)$time == day & DE_genes_response$response == 1)]))

DE_genes_response2_mean <- sapply(c( "Day0", "Day1", "Day3", "Day14"), function(day) rowMeans(assay(
DE_genes_response)[, which(colData(
DE_genes_response)$time == day & DE_genes_response$response == 2)]))


pheatmap(DE_genes_response1_mean, cluster_rows = TRUE,  cluster_cols = FALSE, scale = "row", 
          main = "Heatmap on top 50 DE genes wrt response over time among low responders")

pheatmap(DE_genes_response2_mean, cluster_rows = TRUE,  cluster_cols = FALSE, scale = "row", 
          main = "Heatmap on top 50 DE genes wrt response over time among high responders")

```

### Gene ontology enrichment analysis

```{r, message = FALSE}

DE_gene_response_name <- gost(query = DEgene_pvalue_BH_response$X, organism = "hsapiens", multi_query = FALSE,
                     correction_method = "fdr", user_threshold = 0.01, source = "GO:BP")

head(DE_gene_response_name$result)

table(DE_gene_response_name[["result"]][["source"]])

colnames(DE_gene_response_name$result)

gostplot(DE_gene_response_name, capped = TRUE, interactive = FALSE)

DE_mod_response = DE_gene_response_name$result[,c("query",  "term_id",
                                "term_name", "p_value", "query_size", 
                                "intersection_size", "term_size", 
                                "effective_domain_size", "intersection_size")]

DE_mod_response$"GeneRatio" = paste0(DE_mod_response$intersection_size,  "/", DE_mod_response$query_size)
DE_mod_response$BgRatio = paste0(DE_mod_response$term_size, "/", DE_mod_response$effective_domain_size)
DE_mod_response$"pvalue" <- DE_mod_response$p_value
DE_mod_response$Count <- DE_mod_response$intersection_size
DE_mod_response$Description<- DE_mod_response$term_name
DE_mod_response_enrich  <-  new("enrichResult", result = DE_mod_response)

dotplot(DE_mod_response_enrich , x = 'GeneRatio', color = "pvalue") + ggtitle("GO: Biological process")

```